### ISSUE TITLE
XCP-ng 8.0 host crashed twice since last updates

### ISSUE THREAD

**Issue by user-72632** [xcp-ng/xcp#337]
We have three Supermicro servers (all of which are Citrix HCL approved) that have XCP-ng 8.0 with the latest patches installed. Two of the three servers (identical server make and hardware) have started to crash since applying the latest round of updates. The third, slightly older and different model has not crashed so far. The crashing servers have been running Citrix XenServer as well as all XCP versions flawlessly for years. I am providing server information with crash images from the xen.log. I have full status and crash reports from both occasions and they are at the identical place in the xen/kernel core code. Here is the server specifics:

    XCP-ng 8.0 (latest updates, no extra or testing packages)
    Intel(R) Xeon(R) CPU E5-2697 v3 @ 2.60GHz
    128 GB ECC DDR4 RAM

Here is the xen.log information from the first crash:
![01-27-2020-PANIC-flags](https://user-images.githubusercontent.com/52837806/74673435-2a42bc80-517d-11ea-8207-642c649e37d1.png)
![01-27-2020-RIP-flags](https://user-images.githubusercontent.com/52837806/74673437-2adb5300-517d-11ea-8822-5a1a74300d29.png)
Here is the xen.log information from the second crash:
![02-17-2020-PANIC-flags](https://user-images.githubusercontent.com/52837806/74673449-3595e800-517d-11ea-9552-5692763ef8c2.png)
![02-17-2020-RIP-flags](https://user-images.githubusercontent.com/52837806/74673451-3595e800-517d-11ea-99a6-817ed30bc544.png)

As can be seen in the screen shots, the same routine in the xen/kernel code is being fired on the RIP crash site in irq.c#flush_ready_eoi.

As mentioned, I have full logs in case someone would like to help. I feel that this could be a problem with the updated microcode but not sure. Here is a list of the last updated packages from the yum.log:
![updates-since-crashes-began](https://user-images.githubusercontent.com/52837806/74673492-4a727b80-517d-11ea-8dd9-2f4001ecce80.png)

If I wanted to experiment with backing off some of these updates, what particular ones would be suitable to try first, after the microcode package. Am I at risk of causing more problems by doing this? I have read that there have been problems with the Intel patches in the Xen code, could this be showing up?

Any help would be appreciated as this is rather disconcerting on a production system that has been running flawlessly just prior to these most recent patches. Full disclosure: we are still on the original BIOS for performance reasons so it is possible that the Intel patches/microcode are starting to show problems with the older BIOS. Again, this has not happened just prior to these updates.

Thanks in advance

**Comment by olivierlambert** [comment]
Hi,

As for pro support or community support, we always ask to use the latest firmware and BIOS from your manufacturer. This is a good base to start searching about your issues, before anything else.

**Comment by user-72632** [comment]
Hi @olivierlambert , totally understand where you are coming from in the Support end of it. Just though I would share with the community. Intel has hobbled their platforms so bad that we would take a 50% hit in performance with our workloads if we were to upgrade to the latest BIOS. Add another 30-40% hit to performance that we are already taking on the patches to Xen, Linux, and Windows would leave our servers running as fast as a netbook. Great for Intel's server chip refresh cycle, not so good for small business!

We have good backups from just prior to the last updates, so we will go back down to those until more investigation can be performed. Clearly this is linear page table related and so directly related to Intel's borked patchset!

Thanks for listening

**Comment by olivierlambert** [comment]
Then you should disable Intel security features in Xen but stay on latest BIOS update.

**Comment by user-72632** [comment]
Thanks. We have tried that and empirically found it to be an unusable situation. We downgraded BIOS back to the original. Perhaps we will try again; after we wait for yet another BIOS update from SM fixing Intel's latest problems from January/February that are not even ready yet!

**Comment by user-72632** [comment]
Reopening this issue as I just had another host crash with the identical crash files as the screen shots in this thread show. This time however, the only difference is that I updated the BIOS, BMC, and all LSI RAID FW to the latest up to date versions.

[    0.000000] DMI: Supermicro SYS-2028U-TR4T+/X10DRU-i+, BIOS 3.2a 11/19/2019

**Comment by olivierlambert** [comment]
Can you try on XCP-ng 8.1? (latest Xen)

**Comment by user-72632** [comment]
@olivierlambert these are production servers. Therefore, I have two options right now; (1) as you mentioned, try the Release Candidate (RC) of XCP-ng 8.1, or (2) restore the Host backup from just prior to the last 8.0 yum update (as it was running fine until then) and wait for the final 8.1 release. In either case I would have to wait until the weekend maintenance window. In the meantime, I would like to keep this issue open while these options are explored. If any logs can help your team to solve, please let me know - I have three crash log sets now..

**Comment by olivierlambert** [comment]
I don't think we'll have time to investigate on this until we release 8.1. Maybe someone in the community :)

**Comment by user-72632** [comment]
Understand completely! Maybe someone else will see this as well. Until then, I keep you posted with what I find out based on my options. Thanks again.

**Comment by rushikeshjadhav** [comment]
@user-72632 do you have anything in `/var/crash/`? Share it, in case you have a full stack trace. I'd also suggest to do a `memtest` from any XCP-NG CD to be sure of the RAM sticks.

**Comment by user-72632** [comment]
@rushikeshjadhav thanks for caring. Yes, I have crash dumps for three different occasions on two separate but identical server hardware so I am reasonably confident it is not memory related. Is there any particular file(s) from the crash dump that you would like to see or all 77 of them? I have not debugged stack traces in a long while but it really looks XSO-308-310 related to me.

**Comment by user-72632** [comment]
@rushikeshjadhav I am providing what I believe are the most pertinent logs from this particular crash. If there are others anyone would like to see, please let me know:

[dmesg.kexec.log](https://github.com/xcp-ng/xcp/files/4312800/dmesg.kexec.log)
[dom0.log](https://github.com/xcp-ng/xcp/files/4312801/dom0.log)
[dom0.structures.log](https://github.com/xcp-ng/xcp/files/4312802/dom0.structures.log)
[xen.log](https://github.com/xcp-ng/xcp/files/4312804/xen.log)
[xen.pcpu27.stack.log](https://github.com/xcp-ng/xcp/files/4312805/xen.pcpu27.stack.log)
[xen-crashdump-analyser.log](https://github.com/xcp-ng/xcp/files/4312806/xen-crashdump-analyser.log)

**Comment by rushikeshjadhav** [comment]
@user-72632 Anything besides guest crash logs. I'll go through above ones :)

**Comment by user-72632** [comment]
@rushikeshjadhav these were the only important xen/dom0 /var/crash logs I also have bugtool dumps as well but I do not want to reveal them unless you have a specific one in mind that I can sanitize before sharing.

**Comment by user-72632** [comment]
@rushikeshjadhav in the dom0.log you see the warnings about the page tables. These occurred on two different servers on different physical and virtual CPU sets.
.......
Stack at     7fff367613a8:
	  00007fff367613a0: WARN  paging error for vaddr 0x00007fff367613a8 - level 3 not present
Code:
	  WARN  paging error for vaddr 0x0000555ff7c75fc5 - level 4 not present
Call Trace:
WARN  paging error for vaddr 0x00007fff367613a8 - level 3 not present
........
Let me know if you still want to see any bugtool logs.

**Comment by rushikeshjadhav** [comment]
Right. I browsed through them and looks like its out of memory boundary access of page. Which made xen fault on the page access and kill dom0 with a sysrq. I'll see if xen git has something for this issue in its master branch.

Rather than reverting packages, would it be possible to setup/reinstall a host with base 8.0 iso and not have microcode and xen updates? I hope you have all trusted guests.

Difficulty here is ability to reproduce the issue at will. Do you have any ideas what might be triggering this? Also are your guests Linux/Windows and PV/HVM? Can you switch all of them to one of the mode for few days? 

Generic BIOS change - disable C-states and enable max performance. This may or may not have any impact.

**Comment by user-72632** [comment]
Hope you find something more on xenbits then I did. The only thing I saw related to the Xen calltrace flush_ready_eoi() routine there was[ this commit.](https://xenbits.xen.org/gitweb/?p=xen.git;a=commit;h=53fd9dafeb5a7601bdfb090a631bafbbd3c84fe3)  As far as setting up a fresh 8.0 from ISO, that might be possible over the weekend; however. I was simply going to do a restore from a point just prior to the yum update which would be pristine at that point. All virtual machines on this particular server are Windows 2012 R2 trusted HVM guests. BIOS is enabled with max performance already; however,  I hesitate to disable C-states for this server.

Lastly, and I am not sure it helps, it appears that these crashes occur within the first few hours of a fresh Host reboot (Monday/Tuesday). Interestingly enough, after the warm Host reboot the crash does not occur again that week. I have not found any external issues that may be causing this. During some crashes I see proxy logs with Internet activity at the time of crash and during others there is none - so I have ruled out web-based DoS hacking or misbehaving guests.

**Comment by olivierlambert** [comment]
Xen 4.13 might be an interesting test (bundled with 8.1)

**Comment by user-72632** [comment]
Yes, yes I know, it makes sense what @olivierlambert is saying. I just need to see if I can fit it in the maintenance window this weekend in addition to seeing what @rushikeshjadhav comes up with as well. In any event, I'll update this issue with the outcome.

**Comment by rushikeshjadhav** [comment]
There are multiple EOI related commits in 4.13 tree. 4.11 does not carry the same. Check [link](https://xenbits.xen.org/gitweb/?p=xen.git&a=search&h=refs%2Fheads%2Fstable-4.13&st=commit&s=EOI). Especially the race condition as mentioned [here](https://xenbits.xen.org/gitweb/?p=xen.git;a=commit;h=a646da4a3d96b7666a3ed99521f1e64dbedf19f9). These patches can't be cherry picked as they have dependency on other critical code paths. 

I think 4.13 is better for your systems. With more study, may be a particular xen commit could be identified that introduced this regression / race from last version.

**Comment by olivierlambert** [comment]
@rushikeshjadhav is working for us @user-72632 :wink: So we can't invest too many time (therefore money) for free. That's why it's better than you switch to 8.1 and report.

Alternatively, we could dig with [pro support](https://xcp-ng.com). I hope you understand that our community support is limited on our time needed to support paid users and improve the software :+1: 

Good luck and keep us posted! Your feedback will be interesting for everyone

**Comment by user-72632** [comment]
Vates has went above and beyond on this one - thanks to both of you. I'll be sure to let the community know how this turns out. It may take a while to let the systems run after the install but I will try to do it before the end of the month - which I believe is your target for the final 8.1 release.

**Comment by user-72632** [comment]
Due to other events going on I was unable to find a suitable maintenance window to do the XCP-ng 8.1 RC install. It also does not look like it will happen before the end of the month as well. After following the XCP-ng Forum articles [here](https://xcp-ng.org/forum/topic/2029/xcpng-xen-kernel-crash-fatal-trap-vector-2-nmi) and [here](https://xcp-ng.org/forum/topic/1025/host-crash-guest_4-o-sh_page_fault__guest), I decided to apply the nmi=dom0 workaround in the hopes that I can hold out until the final 8.1 release. I'll leave it up to you if you would like to close this issue or not.

**Comment by rushikeshjadhav** [comment]
I doubt nmi=dom0 would help here as here its a page access fault instead of new interrupt but it certainly won't hurt. You could try reaching xen-devel to understand more about this behavior and if someone can point exact patch that fixed it.

**Comment by user-72632** [comment]
Yes I agree, but you have to try something while you wait or else....

**Comment by user-72632** [comment]
Update: as @rushikeshjadhav expected, it crashed again last night even with the nmi=dom0 boot option. At least this shows that option has nothing to do with this problem. Will once again try to schedule time this weekend for a XCP-ng 8.1 RC install.

**Comment by rushikeshjadhav** [comment]
May I suggest to run the host in `Xen-debug` mode? There is no grub entry for it as of now but you can add following and boot into it. I hope this can elaborate the reason of crash, it won't fix it though.

Use following grub entry, change `root-ohubmf` to your suitable value.

menuentry 'XCP-ng Debug' {
	search --label --set root root-ohubmf
	multiboot2 /boot/xen-debug.gz dom0_mem=8192M,max:8192M watchdog ucode=scan dom0_max_vcpus=1-16 crashkernel=256M,below=4G console=vga vga=mode-0x0311 nmi=dom0
	module2 /boot/vmlinuz-4.19-xen root=LABEL=root-ohubmf ro nolvm hpet=disable xencons=hvc console=hvc0 console=tty0 quiet vga=785 splash plymouth.ignore-serial-consoles
	module2 /boot/initrd-4.19-xen.img
}

**Comment by user-72632** [comment]
Update 03/23/2020: I'm not feeling the love for the 8.1 RC right now.  Over the weekend I tried two different install methods for the XCP-ng 8.1 RC. After backing up with Clonezilla, I first tried to do the easiest way using yum update. On a test VM this method worked well and did not modify much. Unfortunately, in the real world it was a disaster and I won't even begin to bore you with the details other than to say that it was not usable. As the yum approach had warnings about it, I gave the second and recommended method a try as this is the one I normally use. Unfortunately, this meant a complete install and reconfiguration. After going through with the ISO install, all seemed to go well. I rebooted and started to configure the server. One of the first things I do is create my local ext SR on top of a large RAID array. Well, I could not create the local ext SR on my RAID partition. Instead, I received an error that I never have before:

![sr-create-error-rc](https://user-images.githubusercontent.com/52837806/77357860-3e9b4b80-6d1f-11ea-9c07-743d2edb8e98.png)


I looked around briefly but did not see any meaningful errors in the logs. I was under a time crunch so I could not debug the situation and likewise had to terminate the installation attempt. At this point and under a deadline, the only real solution was to restore an image from just before the December updates. All servers are now running Xen version 4.11.1-7.5.1.xcpng8.0, from Fri Aug 30 11:49:34 CEST 2019. This is the version just before the regressions with the interrupt code appeared. I am sticking with this as it is stable for me right now. I need to do more real world machine testing with the RC or final before I decide to try it in production. It looks like changes to incorporate ext4 have somehow introduced regressions. During the next open maintenance window I will try working with this again when I have more time to work on just one server.

**Comment by olivierlambert** [comment]
You are trying to create a SR on a partition that's likely used by the system itself. Are you sure it was `sda1` that was free? IIRC, that's not what I remember of the default layout.

**Comment by user-72632** [comment]
@olivierlambert your going to have to give me a little more credit than that. The system is installed on a software raid based dual 64GB SATADOM setup (md127p1-6). sda1 is for my large local SR and I also have a sda2 for backups. Been using this setup for a very long time and never had any problems with doing this. The sda disk is a LSI/Avago RAID5 setup with all firmware updates installed and I know it was working correctly because I mounted and was using the sda2 partition (ext4) just fine.

**Comment by nagilum99** [comment]
To be fair: It's extremely uncommon to not install OS on your first disks an use 'higher' ones for storage.
So sda* is extremely unlikely to be used for storage.
Usually it would be sda/sdb -> md127... and then sdc(...z) for storage repo.
Maybe that also knocks upgrades off? It's very uncommon, so probably also very untested.

You could also just mount it by hand (mount it to some temp/test-dir), if it's only ext4, to check if that works without any error.

**Comment by user-72632** [comment]
I'm not giving @olivierlambert a hard time believe me.  I'm not sure what environment you are in but with a high end Supermicro server you have two SATADOM sockets on the motherboard. These will show up as sdb and sdc because any AOC (add on card) for RAID will have priority and show up as sda in the boot process. When you install XCP you group sdb and sdc together for software RAID1. Likewise, sda is free to get partitioned anyway you desire, in our use case we have a large local SR and backup partition.

**Comment by olivierlambert** [comment]
"Failure hd" is a bit scarce. Can you check SMlog and xensource log corresponding on the time you tried to create the SR?

**Comment by user-72632** [comment]
Yes, the error message lacks information. When this happened I briefly looked through those logs and did not see anything that jumped out at me. However, I was in a hurry to restore the old system so I did not spend as much time as I would have liked poking around. Now that I have restored a known working system, I will be able to focus on just one server to test this again; hopefully over the weekend. When I do, I will report back with my findings. Hopefully, it was something I did that caused this because I was in rush mode. I'll let you know - thanks again!

**Comment by user-72632** [comment]
Update: I took a production server offline just now to install the official XCP-ng 8.1 ISO from scratch in order to get to the bottom of this issue. Here is the situation:
1. Fresh install of XCP-ng ISO on 2 x 64GB SATADOM in software RAID1.
2. Did not configure SR for VMs during install (always do that late).
3. Install wen fine, on boot up xsconsole is noticeably hosed in the following ways:
   (a) the hostname does not show up in the top right corner of the xsconsole home screen.
   (b) the management interface does not show up on the xsconsole home screen
4. XCP-ng Center cannot connect.
5. I can ssh to it just fine.
6. I am getting Raising exception [150, Failed to initialize XMLRPC connection] in the SMLog
7. Did not even try to create SR yet because I think the network is hosed somehow and therefore the RPC server connection will not work so it is not even worth trying.

I have seen this behavior every time during the installations which is why I thought something was not right. I have 4 x ixgbe based Intel X540-AT2. These don't seem to be the problem as the ixgbe driver has not changed from 8.0. Something strange is going on with the networking that is causing some fundamental problem with a lot of other things.

I have never had trouble like this doing an install just this way many times. Could someone help with this so we can find what is going on. The system is running now and I have not touched it. I can supply logs to help troubleshoot.

**Comment by olivierlambert** [comment]
Do you have an Nvidia GPU?

Try a `yum install gpumon` and reboot.

**Comment by user-72632** [comment]
Wow, that is weird, yes I do have a K2200 installed. I'll give it a go and let you know.

**Comment by olivierlambert** [comment]
Yeah, imagine how hard it was to diagnose this… kudos to @stormi

**Comment by user-72632** [comment]
Okay, I installed gpumon and rebooted. I get the root@hostname in xsconsole but still no management network configured on the home screen. The SMLog now looks like it is filling up with info now and that never happened. Also, the NICS are not showing in xsconsole so I think there is still something wrong.

**Comment by stormi** [comment]
On one host we also had to do an emergency network reset.

**Comment by user-72632** [comment]
Okay, I can try that but then it will dhcp me ....

**Comment by user-72632** [comment]
Wow, that looks good! I will continue on with my testing now to create SR etc, and let you know if I have any more issues. Also, why was the GPU monitoring daemon left out? And, will another ISO be created that has it already installed?

**Comment by user-72632** [comment]
@olivierlambert and @stormi, it looks good so far, SR creation, bonding, GPU passthrough, SCSI card passthrough, all configured like a dream. The ext4 SR create was nice it went fast not like before with ext3 taking forever to create superblocks. I am running this one server in production now with heavy use under a virtualized CAD workload so I will see if my original Xen crash problem has been solved now that it is using 4.13 instead of 4.11. Fingers crossed.

I just want to say, I hope this is not Citrix up to their old tricks by leaving an important piece out of the puzzle. Nice catch by @stormi ! I will update this post in about a week to give the server a chance to run. Do you think this might have some bearing on the other software RAID1 restore issue as well?

Thanks again!

**Comment by user-72632** [comment]
@stormi and @borzel I noticed that XCP-ng Center no longer shows the DVD or removable drives in the tree view for the new XCP 8.1 install as shown here:
![xcp-issue](https://user-images.githubusercontent.com/52837806/78504296-3dfcaf00-773a-11ea-9003-9440d7e43e19.png)
I don't think it is XCP-ng Center's problem but thought I would mention it. It looks like Center is just reporting what it sees but that the XAPI is not seeing it. Looking at "xe vdi-list" on the 8.1 install does not show the required devices that the 8.0 install has. This may be another side-effect of the gpumon manual install. I won't know until next weekend when I do the final ISO install if the Xen crash issues do not appear this week.

**Comment by stormi** [comment]
Thanks for reporting this. I've mentioned this possiblity in https://github.com/xcp-ng/xcp/wiki/XCP-ng-8.1-Known-Issues based on your feedback.

**Comment by user-72632** [comment]
Your welcome. FYI: out of necessity last night (8.0 crashed hard again), I installed the latest 8.1 ISO you put together with all the fixes and it went on smooth. I don't want to jinx it but it seems to be very stable and the nVidia GPU is getting quite a workout. All the DVD and removable devices showed up as normal and the SR was created easily. All PCI pass through devices are working well also. I'll let that run this week and report back over the weekend to let you know if I have any other issues.

Thanks again and BTW, I saw your input to the XAPI folks over there and that looks encouraging with them working on an alternative. I'll watch that space for any further developments as it interest me - well done! I still need to try out the restore issue from #357 but that will also have to wait until later.
